### PR TITLE
convert trigger events from ioctl syscall to a uprobe

### PR DESCRIPTION
**Pull request by AsafEitani**
## Initial Checklist

- [x] There is an issue describing the need for this PR.
- [x] Git log contains summary of the change.
- [x] Git log contains motivation and context of the change.
- [x] If part of an EPIC, PR git log contains EPIC number.
- [x] If part of an EPIC, PR was added to EPIC description.

## Description (git log)

commit 17b4c94551824780e6213d7bf7eacc82506324f6 (HEAD -> uprobe, origin/uprobe)
Author: AsafEitani <eitaniasaf@gmail.com>
Date:   Sat Jul 30 15:57:10 2022 +0300

    events: add context of triggering event to the triggered event
    
    Retains the event context of the triggering event (e.g. hooked_syscall gets the context of the insmod that preceded it)

commit c62d19cfb0055ea1adac9abf4fdf4bb9e2fe3fdb
Author: AsafEitani <eitaniasaf@gmail.com>
Date:   Mon Jun 6 16:00:37 2022 +0300

    events: convert trigger events to uprobe
    
    Change the way tracee-ebpf is triggering events.
    Solves #2019

Fixes: #2019

## Type of change

- [ ] Bug fix (non-breaking change fixing an issue, preferable).
- [ ] Quick fix (minor non-breaking change requiring no issue, use with care)
- [x] Code refactor (code improvement and/or code removal)
- [x] New feature (non-breaking change adding functionality).
- [ ] Breaking change (cause existing functionality not to work as expected).

## How Has This Been Tested?

Tested both on x86 and ARM.

Reproduce the test by running:

`tracee-ebpf -t e=hooked_syscalls,hooked_seq_ops` and load a rootkit like diamorphine that hooks syscalls or seq ops.

## Final Checklist:

Pick "Bug Fix" or "Feature", delete the other and mark appropriate checks.

- [x] I have made corresponding changes to the documentation.
- [x] My code follows the style guidelines (C and Go) of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented all functions/methods created explaining what they do.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix, or feature, is effective.
- [x] New and existing unit tests pass locally with my changes.
- [x] Any dependent changes have been merged and published before.

## Git Log Checklist:

My commits logs have:

- [x] Subject starts with "subsystem|file: description".
- [x] Do not end the subject line with a period.
- [x] Limit the subject line to 50 characters.
- [x] Separate subject from body with a blank line.
- [x] Use the imperative mood in the subject line.
- [x] Wrap the body at 72 characters.
- [x] Use the body to explain what and why instead of how.

### CODE REVIEW

File: pkg/ebpf/events_processor.go
Comment (NDStrahilevitz) [inline pkg/ebpf/events_processor.go:368]: This map is shared between two separate goroutines, it needs a mutex.

File: pkg/ebpf/tracee.go
Comment (NDStrahilevitz) [inline pkg/ebpf/tracee.go:915]: These probes should be integrated into the probes package.

File: pkg/ebpf/tracee.go
Comment (NDStrahilevitz) [inline pkg/ebpf/tracee.go:919]: ditto

File: pkg/ebpf/tracee.go
Comment (NDStrahilevitz) [inline pkg/ebpf/tracee.go:1160]: This logic can be used in probes.go as part of integration.

File: pkg/events/derive/derive.go
Comment (NDStrahilevitz) [inline pkg/events/derive/derive.go:15]: ditto on the mutex.
I'm also not sure if this is the best place to have the map, Maybe it should get it's own pacakge like most other caches, and then it can be managed there, with memory safety and all. It can be given as a dependency to the deriving functions.

**Review by NDStrahilevitz — CHANGES_REQUESTED** [review]
I like this for being cleaner than the previous solution, but:
1. Uprobes aren't integrated into the existing probes code.
2. The invoking map, imo, doesn't belong in the deriving code, I believe this is why a mutex was also more easily missed.

**Comment by rafaeldtinoco** [comment]
Please don't forget to fully update our PR template. I self-requested review because I want to follow the uprobe work if possible! Thanks.

File: pkg/ebpf/tracee.go
Comment (rafaeldtinoco) [inline pkg/ebpf/tracee.go:1144]: Ingenious =D

File: pkg/ebpf/tracee.go
Comment (rafaeldtinoco) [reply pkg/ebpf/tracee.go:1172]: https://github.com/rafaeldtinoco/tracee/commit/19f112bdc5d6f7694a61025980edeebcfcbd2fba

Since I recently touched that code, I created the needed change if you want to pick that commit and merge into your PR and push it.

File: pkg/ebpf/tracee.go
Comment (rafaeldtinoco) [reply pkg/ebpf/tracee.go]: https://github.com/aquasecurity/tracee/pull/2031#discussion_r936178427

**Comment by rafaeldtinoco** [comment]
commit 23bbc9b2 needs "rewording" (git rebase -i -> reword) for line breaks.

File: pkg/events/derive/derive.go
Comment (rafaeldtinoco) [inline pkg/events/derive/derive.go:72]: Shouldn't we just add an argument to this function (isTriggered true/false) instead of duplicating the previous function ?

File: pkg/events/derive/derive.go
Comment (rafaeldtinoco) [inline pkg/events/derive/derive.go:96]: Same for this one.

File: pkg/events/derive/derive.go
Comment (NDStrahilevitz) [reply pkg/events/derive/derive.go:72]: I think it would make more sense to do something similar to the `deriveArgsFunction` and pass an `eventBuilder` function. Then we reduce to a single function receiving both (probably the existing `singleEventDeriveFunc`).

File: pkg/events/derive/derive.go
Comment (NDStrahilevitz) [inline pkg/events/derive/derive.go:86]: Maybe something simpler like `withInvokingContext` similar to how golang contexts do it.

File: pkg/events/derive/derive.go
Comment (NDStrahilevitz) [inline pkg/events/derive/derive.go:104]: This can be left unexported I believe so `baseEvent`.

File: pkg/events/derive/derive.go
Comment (NDStrahilevitz) [inline pkg/events/derive/derive.go:16]: I think we shouldn't leave this as a global variable, and now we have a global mutex too.
It could very simply be fit into a struct like the `procinfo` cache.

**Review by NDStrahilevitz — CHANGES_REQUESTED** [review]

File: pkg/ebpf/c/tracee.bpf.c
Comment (yanivagman) [inline pkg/ebpf/c/tracee.bpf.c]: Where does this 16 comes from? Is it something specific to ARM64?
If so, better if we add if defined(bpf_target_arm64)

File: pkg/ebpf/c/tracee.bpf.c
Comment (yanivagman) [inline pkg/ebpf/c/tracee.bpf.c:86]: This can be passed as an argument to the ioctl, which also has the advantage of removing the need to sync it with userspace struct size.

**Review by yanivagman — DISMISSED** [review]
That's a good change. Need to check that it works well when tracee is deployed inside a container as well, so we won't have surprises on release.

File: pkg/ebpf/events_processor.go
Comment (rafaeldtinoco) [reply pkg/ebpf/events_processor.go:374]: Yes, I thought the same in previous review, forgot to mention.

File: pkg/ebpf/events_processor.go
Comment (rafaeldtinoco) [reply pkg/ebpf/events_processor.go:375]: Also, all the SeqOps should contain "NetSeqOps" as there are multiple seqops structures in kernel (if you agree).

**Review by rafaeldtinoco — COMMENTED** [review]
LGTM after Yaniv's points are addressed.

File: pkg/ebpf/c/tracee.bpf.c
Comment (AsafEitani) [reply pkg/ebpf/c/tracee.bpf.c:3114]: Can we do this in a different PR? this will take some testing

File: pkg/ebpf/c/tracee.bpf.c
Comment (AsafEitani) [reply pkg/ebpf/c/tracee.bpf.c:86]: Can we do this in a different PR? this will take some testing

File: pkg/ebpf/probes/probes.go
Comment (NDStrahilevitz) [inline pkg/ebpf/probes/probes.go:281]: Idea for the future, I think we can have this struct have a `Trigger` method of it's own, then instead of manually building the `symbolName` we can also generate it (of course matching to an existing bpf uprobe code).

This could (possibly) also help with a more concrete method of event invocation, an invoked event from a trigger would have its context stored "by definition" through this `Trigger`.

Comment (rafaeldtinoco) [reply pkg/ebpf/probes/probes.go:281]: I would wait for more uprobe usage needs (specially those unrelated to tracee binary itself, like these ones) and then define something else if needed.

File: pkg/ebpf/tracee.go
Comment (rafaeldtinoco) [inline pkg/ebpf/tracee.go:1112]: Alright, let's be honest here.. the previous function `invokeIoctlTriggeredEvents()` already existed as a function for `*Tracee` so I'm willing to accept/agree having these functions here for now, but this should be **temporary** as the functions are unrelated to `*Tracee` and related to specific events.

If we compare all other `(*Tracee).Functions` we have "initBPF", "getOptionsConfig", "generateInitValues", ... and they're all generic to functioning of Tracee (not to specific events), or they should be.

We should probably create an interface for anything relying in similar approach(es) (anything that needs a trigger + derived events + re-trigger, based in uprobes and things like that). @NDStrahilevitz for the record here (as he is our refactoring guy =D).

I'm willing to move forward as we're basically replacing the ioctl() mechanism for uprobe, but I think it needs refactor in the next version (and an issue opened).

File: pkg/ebpf/c/tracee.bpf.c
Comment (rafaeldtinoco) [reply pkg/ebpf/c/tracee.bpf.c:86]: Mind opening an issue and putting a TODO in the code so we don't forget ?

File: pkg/ebpf/c/tracee.bpf.c
Comment (rafaeldtinoco) [reply pkg/ebpf/c/tracee.bpf.c:3111]: Mind opening an issue and putting a TODO in the code so we don't forget ?

**Review by rafaeldtinoco — COMMENTED** [review]
@AsafEitani

I believe we're half way through for the "final solution", probably with a "functional enough" code for this release. I'm willing to merge (for 0.8.1) as I know its important for now... but there are some issues to be opened (and TODOs to be added to the code) and addressed for 0.8.2 (related to uprobe arguments and the uprobe symbol/functions placed somewhere else). 

I'll give a second look at the 5.4 verifier changes patches before merging, and I would like to ask @NDStrahilevitz your input here as well, so we can merge this asap (9/9/22 ?) for the 0.8.1 release.

File: pkg/ebpf/c/tracee.bpf.c
Comment (rafaeldtinoco) [reply pkg/ebpf/c/tracee.bpf.c]: Yep, is this a calling convention for the uprobe eBPF programs contexts in both arches ? Did you find it by increasing the offset or is it documented somewhere ? A comment here would be nice as well.

File: pkg/ebpf/tracee.go
Comment (NDStrahilevitz) [reply pkg/ebpf/tracee.go:1112]: Yep, please add a TODO to move away from this.

File: pkg/ebpf/tracee.go
Comment (NDStrahilevitz) [inline pkg/ebpf/tracee.go:1143]: Main challenge for moving the triggers might be defining the "probe arguments" more generically, but we'll tackle that when refactoring this.

File: pkg/events/derive/hooked_seq_ops.go
Comment (NDStrahilevitz) [inline pkg/events/derive/hooked_seq_ops.go:48]: Could you make this indexing more concise/human-readable (maybe a "translation" function)

**Review by NDStrahilevitz — COMMENTED** [review]
So, for this part of this change "package" (including #2049) i'm basically ok with this one, my main gripe overall is the way we retain context, which i've addressed over there.

So for this one I think we're good on merging for now, just add the comments and TODOs requested and open an issue for refactoring appropriately in 0.8.2, mainly on where and how we trigger the probes, this could be solved overall in the context of an "event invocation" package, similar to how we have a package for event derivation.

File: pkg/ebpf/c/tracee.bpf.c
Comment (AsafEitani) [reply pkg/ebpf/c/tracee.bpf.c]: So it is documented but I had to do some testing to figure which argument goes where.
This is the documentation:
https://go.googlesource.com/go/+/refs/heads/dev.regabi/src/cmd/compile/internal-abi.md

I'll add a comment

File: pkg/ebpf/c/tracee.bpf.c
Comment (AsafEitani) [reply pkg/ebpf/c/tracee.bpf.c:3111]: #2055 
Will add a TODO in the code

File: pkg/ebpf/tracee.go
Comment (AsafEitani) [reply pkg/ebpf/tracee.go:1112]: Sure I agree, I'll add a TODO and add that to #2055

File: pkg/ebpf/tracee.go
Comment (rafaeldtinoco) [reply pkg/ebpf/tracee.go:1147]: https://github.com/aquasecurity/tracee/issues/2055

File: pkg/ebpf/c/tracee.bpf.c
Comment (rafaeldtinoco) [reply pkg/ebpf/c/tracee.bpf.c]: Thanks for the pointers for the golang calling convention (as a comment).

File: pkg/ebpf/c/tracee.bpf.c
Comment (rafaeldtinoco) [inline pkg/ebpf/c/tracee.bpf.c:3105]: I left a suggestion for changes at: https://github.com/aquasecurity/tracee/pull/2049#discussion_r941866347 after this was merged.

**Pull request merged**
